packages/tpetra/core/src/Tpetra_CrsMatrix_def.hpp

```diff
@@ -1943,37 +1943,46 @@ namespace Tpetra {
 #endif // HAVE_TPETRA_DEBUG
 
     RowInfo rowInfo = graph.getRowInfo (lclRow);
-    const size_t curNumEnt = rowInfo.numEntries;
-    const size_t newNumEnt = curNumEnt + numEntriesToAdd;
-    if (newNumEnt > rowInfo.allocSize) {
-      TEUCHOS_TEST_FOR_EXCEPTION_CLASS_FUNC
-        (this->getProfileType () == StaticProfile, std::runtime_error,
-         "New indices exceed statically allocated graph structure.");
-      // This must be a nonconst reference, since we'll reallocate.
-      Teuchos::Array<IST>& curVals = this->values2D_[lclRow];
-      // Make space for the new matrix entries.
-      // Teuchos::ArrayRCP::resize automatically copies over values on
-      // reallocation.
-      graph.lclInds2D_[rowInfo.localRow].resize (newNumEnt);
-      curVals.resize (newNumEnt);
-      rowInfo.allocSize = newNumEnt; // give rowInfo updated allocSize
-    }
-    typename crs_graph_type::SLocalGlobalViews indsView;
-    indsView.linds = indices;
-
-    Teuchos::ArrayView<IST> valsView = this->getViewNonConst (rowInfo);
-    Teuchos::ArrayView<const IST> valsIn =
-      Teuchos::av_reinterpret_cast<const IST> (values);
-    this->insertIndicesAndValues (graph, rowInfo, indsView, valsView,
-                                  valsIn, LocalIndices, LocalIndices);
+
+    if (this->getProfileType() == StaticProfile)
+    {
+      Teuchos::ArrayView<IST> valsView = this->getViewNonConst(rowInfo);
+      auto fun = [&](size_t const k, size_t const /*start*/, size_t const offset) {
+                   valsView[offset] += values[k]; };
+      graph.insertLocalIndicesImpl(lclRow, indices, fun);
```

Comment:
Since it's necessary for the graph to take `fun` as an `std::function`, and since the graph won't keep the `std::function`, we could create the `std::function` explicitly by reference.  This would trigger small object optimizations for `std::function`.  That is, it would end up just storing a single reference, instead of various ArrayView stuff (that would likely bring it over the small object optimization limit).  Here's how to do it:
```c++
// Prefer std::cref.  You can't use it here, since
// ArrayView<IST>::operator[] is nonconst.
std::function<void(const size_t, const size_t, const size_t)> fun2 (std::ref (fun));
// You can't inline the declaration of fun inside fun2 above,
// since that would make the lambda const and would require
// std::ref(const T&&), which is deleted.
graph.insertLocalIndicesImpl (lclRow, indices, fun2);
```

Reply:
done.